src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java

```diff
@@ -0,0 +1,268 @@
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("Started")
+                .willSetStateTo("1")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBody("{\n" +
+                            "  \"token\": \"super-secret-token\",\n" +
+                            // This token will go stale at the soonest allowed time but will no be expired for the duration of the test
+                            "  \"expires_at\": \"" + printDate(new Date(System.currentTimeMillis() + Duration.ofMinutes(10).toMillis())) + "\"" + // 2019-08-10T05:54:58Z
+                            "}")));
+
+        // Force an error to test fallback refreshing from agent
+        githubApi.stubFor(
+            post(urlEqualTo("/app/installations/654321/access_tokens"))
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("1")
+                .willSetStateTo("2")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withStatus(404)
+                        .withStatusMessage("404 Not Found")
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBody("{\"message\": \"File not found\"}")));
+
+        // Force an error to test fallback to returning unexpired token
+        githubApi.stubFor(
+            post(urlEqualTo("/app/installations/654321/access_tokens"))
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("1")
+                .willSetStateTo("2")
+                .withRequestBody(
+                    equalToJson(
+                        "{\"permissions\":{\"pull_requests\":\"write\",\"metadata\":\"read\",\"checks\":\"write\",\"contents\":\"read\"}}", true, false))
+                .willReturn(
+                    aResponse()
+                        .withStatus(404)
+                        .withStatusMessage("404 Not Found")
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBody("{\"message\": \"File not found\"}")));
+
+        // Force an error to test fallback refreshing from agent
+        githubApi.stubFor(
+            post(urlEqualTo("/app/installations/654321/access_tokens"))
+                .inScenario(scenarioName)
+                .whenScenarioStateIs("2")
+                .willSetStateTo("3")
```

Comment:
Oh neat, did not know about this feature of WireMock.

Reply:
Absolutely vital. 😄